crates/steel-parser/src/parser.rs

```diff
@@ -1350,79 +1380,100 @@ impl ASTLowerPass {
                             let value = std::mem::replace(value, List::new(vec![]));
 
                             *expr = match &a.syn.ty {
-                                TokenType::If => parse_if(value.args.into_iter(), a.syn),
+                                TokenType::If => {
+                                    parse_if(value.args_proper(TokenType::If)?.into_iter(), a.syn)
+                                }
```

Comment:
These help with things like `(let ((x 1)) . 1)`. I'm not sure this properly catches all possible mistakes, since lowering is a bit all over the place.

(Also, I'm assuming that no special form allows dots, but I haven't checked)

Reply:
The lowering is all over the place for a few reasons, and in general I'd like to clean that up - but the reason was that originally I parsed directly into the core forms. Once macros were added, I instead parsed into lists before lowering to the core forms - and also when dealing with kernel level macros, you have to go back and forth sometimes.

I think this is fine though - I will take a holistic pass and see where all the lowering is happening